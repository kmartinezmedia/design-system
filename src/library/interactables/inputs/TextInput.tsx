import React, { useState, ReactNode } from 'react';
import {
  View,
  TextInput as RNTextInput,
  NativeSyntheticEvent,
  TextInputFocusEventData,
  TextInputProps as RNTextInputProps,
  StyleSheet,
  Keyboard,
} from 'react-native';
import {
  useTheme,
  elevation,
  border,
  spacing,
  fontSize,
  sizing,
} from '@designSystem/theme';
import { Caption, Label, VStack } from '@designSystem/primitives';
import { InputVariant } from '@designSystem/types';
import { useInputVariant } from '@designSystem/hooks';

export interface TextInputProps extends RNTextInputProps {
  testID?: string;
  children?: (p: TextInputProps) => ReactNode;
  variant?: InputVariant;
  width?: '100%' | 'auto';
  label?: string;
  description?: React.ReactNode;
  icon?: React.ReactNode;
  append?: boolean;
  error?: string;
}

export const TextInput = React.forwardRef<RNTextInput, TextInputProps>(
  (
    {
      variant = 'default',
      width = '100%',
      autoCapitalize = 'sentences',
      returnKeyType = 'next',
      textContentType = 'none',
      onSubmitEditing = Keyboard.dismiss,
      onFocus,
      onBlur,
      value,
      label,
      description,
      icon,
      children,
      append,
      error,
      testID,
      ...inputProps
    },
    ref,
  ) => {
    const { colors } = useTheme();
    const [focused, setFocused] = useState(false);
    const [borderStyles, inputStyles, labelColor] = useInputVariant(
      variant,
      focused,
    );

    const props = {
      testID,
      style: [styles.input, inputStyles, append && styles.append],
      value,
      autoCapitalize,
      returnKeyType,
      autoCorrect: false,
      placeholderTextColor: colors.onBackground.muted,
      onFocus: (e: NativeSyntheticEvent<TextInputFocusEventData>) => {
        onFocus && onFocus(e);
        setFocused(true);
      },
      onBlur: (e: NativeSyntheticEvent<TextInputFocusEventData>) => {
        onBlur && onBlur(e);
        setFocused(false);
      },
      onSubmitEditing,
      ...inputProps,
    };

    return (
      <VStack spacing={error || description ? { bottom: 2 } : 0}>
        {!!label && <Label color={labelColor}>{label}</Label>}
        <View
          style={[
            {
              width,
              marginBottom: description || error ? spacing[1] : spacing[4],
            },
            styles.border,
            append && styles.append,
            borderStyles,
          ]}>
          {children ? (
            children(props)
          ) : (
            <RNTextInput
              {...props}
              textContentType={textContentType}
              keyboardAppearance="light"
              ref={ref}
            />
          )}
          {!!icon && <View style={styles.icon}>{icon}</View>}
        </View>
        {description ||
          (!!error && <Caption color="negative">{error}</Caption>)}
      </VStack>
    );
  },
);

const styles = StyleSheet.create({
  // Note: we do this nasty hack for brownfield consumer integration
  // for some reason consumer iOS was animating the padding property, so
  // we're faking padding by using a background colored border as spacing on the text
  // input - then adding a wrapper element to show the actual border outline
  border: {
    borderRadius: border.inputRadius,
    borderWidth: border.width,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
  },

  append: {
    borderTopLeftRadius: 0,
    borderBottomLeftRadius: 0,
  },

  input: {
    fontSize: fontSize.body,
    width: '100%',
    borderStyle: 'solid',
    borderRadius: border.inputRadius - border.width,
    paddingHorizontal: spacing[3],
    paddingVertical: 0,
    flex: 1,
    height: sizing.input - 2 * border.width,
  },

  icon: {
    paddingRight: spacing[3],
    zIndex: elevation.layer2,
  },
});
