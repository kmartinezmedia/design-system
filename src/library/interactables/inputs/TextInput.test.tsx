import React from 'react';
import { create, act, ReactTestRenderer } from 'react-test-renderer';
import { TextInput } from './TextInput';

describe('TextInput', () => {
  it('default', () => {
    let component: ReactTestRenderer;
    act(() => {
      component = create(
        <>
          <TextInput />
          <TextInput variant="positive" />
          <TextInput variant="negative" error="you messed up" />
        </>,
      );
    });
    expect(component!.toJSON()).toMatchSnapshot();
  });
});
