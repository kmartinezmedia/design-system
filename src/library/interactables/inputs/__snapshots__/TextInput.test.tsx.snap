// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput default 1`] = `
Array [
  <View
    style={
      Array [
        Object {
          "alignItems": "stretch",
          "alignSelf": "auto",
          "flexDirection": "column",
          "flexGrow": undefined,
          "flexShrink": undefined,
          "flexWrap": "nowrap",
          "height": undefined,
          "justifyContent": "flex-start",
          "width": undefined,
        },
        undefined,
        undefined,
        undefined,
        Object {},
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "marginBottom": 16,
            "width": "100%",
          },
          Object {
            "alignItems": "center",
            "borderRadius": 4,
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "center",
          },
          undefined,
          Object {
            "backgroundColor": "#FFFFFF",
            "borderColor": "#F1F3F5",
          },
        ]
      }
    >
      <TextInput
        allowFontScaling={true}
        autoCapitalize="sentences"
        autoCorrect={false}
        keyboardAppearance="light"
        onBlur={[Function]}
        onFocus={[Function]}
        onSubmitEditing={[Function]}
        placeholderTextColor="#587088"
        rejectResponderTermination={true}
        returnKeyType="next"
        style={
          Array [
            Object {
              "borderRadius": 3,
              "borderStyle": "solid",
              "flex": 1,
              "fontSize": 16,
              "height": 46,
              "paddingHorizontal": 12,
              "paddingVertical": 0,
              "width": "100%",
            },
            Object {
              "backgroundColor": "#FFFFFF",
              "borderColor": "#FFFFFF",
              "color": "#071422",
            },
            undefined,
          ]
        }
        textContentType="none"
        underlineColorAndroid="transparent"
      />
    </View>
  </View>,
  <View
    style={
      Array [
        Object {
          "alignItems": "stretch",
          "alignSelf": "auto",
          "flexDirection": "column",
          "flexGrow": undefined,
          "flexShrink": undefined,
          "flexWrap": "nowrap",
          "height": undefined,
          "justifyContent": "flex-start",
          "width": undefined,
        },
        undefined,
        undefined,
        undefined,
        Object {},
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "marginBottom": 16,
            "width": "100%",
          },
          Object {
            "alignItems": "center",
            "borderRadius": 4,
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "center",
          },
          undefined,
          Object {
            "backgroundColor": "#FFFFFF",
            "borderColor": "#00A45F",
          },
        ]
      }
    >
      <TextInput
        allowFontScaling={true}
        autoCapitalize="sentences"
        autoCorrect={false}
        keyboardAppearance="light"
        onBlur={[Function]}
        onFocus={[Function]}
        onSubmitEditing={[Function]}
        placeholderTextColor="#587088"
        rejectResponderTermination={true}
        returnKeyType="next"
        style={
          Array [
            Object {
              "borderRadius": 3,
              "borderStyle": "solid",
              "flex": 1,
              "fontSize": 16,
              "height": 46,
              "paddingHorizontal": 12,
              "paddingVertical": 0,
              "width": "100%",
            },
            Object {
              "backgroundColor": "#FFFFFF",
              "borderColor": "#FFFFFF",
              "color": "#00A45F",
            },
            undefined,
          ]
        }
        textContentType="none"
        underlineColorAndroid="transparent"
      />
    </View>
  </View>,
  <View
    style={
      Array [
        Object {
          "alignItems": "stretch",
          "alignSelf": "auto",
          "flexDirection": "column",
          "flexGrow": undefined,
          "flexShrink": undefined,
          "flexWrap": "nowrap",
          "height": undefined,
          "justifyContent": "flex-start",
          "width": undefined,
        },
        Object {
          "paddingBottom": 8,
        },
        undefined,
        undefined,
        Object {},
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "marginBottom": 4,
            "width": "100%",
          },
          Object {
            "alignItems": "center",
            "borderRadius": 4,
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "center",
          },
          undefined,
          Object {
            "backgroundColor": "#FFFFFF",
            "borderColor": "#F1F3F5",
          },
        ]
      }
    >
      <TextInput
        allowFontScaling={true}
        autoCapitalize="sentences"
        autoCorrect={false}
        keyboardAppearance="light"
        onBlur={[Function]}
        onFocus={[Function]}
        onSubmitEditing={[Function]}
        placeholderTextColor="#587088"
        rejectResponderTermination={true}
        returnKeyType="next"
        style={
          Array [
            Object {
              "borderRadius": 3,
              "borderStyle": "solid",
              "flex": 1,
              "fontSize": 16,
              "height": 46,
              "paddingHorizontal": 12,
              "paddingVertical": 0,
              "width": "100%",
            },
            Object {
              "backgroundColor": "#FFFFFF",
              "borderColor": "#FFFFFF",
              "color": "#071422",
            },
            undefined,
          ]
        }
        textContentType="none"
        underlineColorAndroid="transparent"
      />
    </View>
    <Text
      style={
        Array [
          Object {
            "fontFamily": "Graphik-Regular",
            "fontSize": 13,
            "fontWeight": "400",
            "lineHeight": 16,
          },
          Object {
            "paddingBottom": 4,
          },
          Object {
            "color": "#E5212E",
            "lineHeight": 16,
            "overflow": "visible",
            "textAlign": undefined,
          },
          undefined,
        ]
      }
    >
      you messed up
    </Text>
  </View>,
]
`;
